storage-provider/server/src/main.rs

```diff
@@ -390,63 +367,77 @@ impl Server {
-
-        cancellation_token.cancel();
-        tracing::info!("sent shutdown signal");
-
-        // Wait for the tasks to finish
-        let (upload_result, rpc_task, pipeline_task, p2p_task, indexer_task) = tokio::join!(
-            storage_task,
-            rpc_task,
-            pipeline_task,
-            p2p_task,
-            indexer_task,
+        let mut tasks = JoinSet::new();
+        tasks.spawn(
+            start_rpc_server(rpc_state, cancellation_token.child_token())
+                .map(|result| ("RPC", result.map_err(ServerError::from))),
         );
-
-        // Inspect and log errors
-        let (upload_result, rpc_task, pipeline_task, p2p_task, indexer_task) = inspect_and_log_nested_errors!(
-            upload_result,
-            rpc_task,
-            pipeline_task,
-            p2p_task,
-            indexer_task,
+        tasks.spawn(
+            p2p::Worker::new(p2p_args)?
+                .run(cancellation_token.child_token())
+                .map(|result| ("P2P", result.map_err(ServerError::from))),
         );
+        tasks.spawn(
+            start_upload_server(Arc::new(storage_state), cancellation_token.child_token())
+                .map(|result| ("HTTP Storage", result.map_err(ServerError::from))),
+        );
+        tasks.spawn(
+            start_pipeline(
+                Arc::new(pipeline_state),
+                pipeline_rx,
+                cancellation_token.child_token(),
+            )
+            .map(|result| ("Pipeline", result.map_err(ServerError::from))),
+        );
+        tasks.spawn(
+            start_indexer(indexer_state, indexer_rx, cancellation_token.child_token())
+                .map(|result| ("Indexer", result.map_err(ServerError::from))),
+        );
+        tracing::info!("Successfully launched all sub-services, ready for work!");
 
-        // Exit with error
-        upload_result??;
-        rpc_task??;
-        pipeline_task??;
-        p2p_task??;
-        indexer_task??;
-
-        Ok(())
+        // Keep the first error around as the "canonical return",
+        // since we can't return multiple values
+        let mut error = None;
+        loop {
+            tokio::select! {
+                result = tasks.join_next() => {
+                    match result {
+                        Some(Ok((task_name, task_result))) => {
+                            match task_result {
+                                Ok(()) => tracing::info!("{task_name} finished successfully!"),
+                                Err(err) => {
+                                    tracing::error!("{task_name} finished with error: {err}");
+                                    tracing::error!("Cancelling remaining tasks...");
+                                    if error.is_none() {
+                                        error = Some(err);
+                                    }
+                                    cancellation_token.cancel();
```

Comment:
Which branches of the match statement are executed if one task fails and others are cancelled?
Can it be like that, that two fail, other ones are cancelled but we're saving an error only from the one?

Reply:
We're only *saving* the first error, yes. But we're logging them all.

If the tasks in the set finish, either successfully or with an error they should be returned by `join_next`. 
Thus, they should all be handled by the respective branches.